// @base44/vite-plugin/dist/injections/page-height-bridge.js
// page-height-bridge — postMessage protocol the iframe exposes to its parent.
// Inert until the parent posts a message; no observers, rewrites, or timers
// fire on their own.
//
//   parent → child  { type: "freeze-vh-units", referenceVhBase?: number }
//     Rewrites every `vh` in <style> + CSSOM to fixed `px` (kills the
//     vh ↔ auto-resize feedback loop). Idempotent; covers HMR-added
//     styles too. Fire-and-forget.
//
//   parent → child  { type: "measure-page-height", settleMs?: number }
//     After `settleMs` (default 2000) posts the page's scrollHeight back
//     as `{ type: "page-height-measured", height }` to the requester.
const FALLBACK_VHBASE = 900;
const MIN_VHBASE = 400;
const DEFAULT_SETTLE_MS = 2000;
const NEUTRALIZE_DEBOUNCE_MS = 16;
const noop = () => { };
let started = false;
/**
 * Installs the parent-driven message listener. Returns a teardown that
 * removes the listener, disconnects any observers attached as a result of
 * `freeze-vh-units`, and clears any pending response timer. Returns a
 * no-op when bailed (already started, SSR, top window).
 */
export function setupPageHeightBridge() {
    if (started)
        return noop;
    if (typeof window === "undefined")
        return noop;
    if (window.self === window.top)
        return noop;
    started = true;
    const controller = createPageHeightBridgeController();
    const onMessage = (event) => {
        const data = (event.data ?? null);
        if (!data || typeof data !== "object")
            return;
        switch (data.type) {
            case "freeze-vh-units": {
                const override = typeof data.referenceVhBase === "number" ? data.referenceVhBase : undefined;
                controller.freezeVhUnits(override);
                return;
            }
            case "measure-page-height": {
                const settleMs = typeof data.settleMs === "number" ? data.settleMs : DEFAULT_SETTLE_MS;
                const origin = event.origin && event.origin !== "null" ? event.origin : "*";
                controller.measurePageHeight(origin, settleMs);
                return;
            }
        }
    };
    window.addEventListener("message", onMessage);
    let torn = false;
    return () => {
        if (torn)
            return;
        torn = true;
        started = false;
        window.removeEventListener("message", onMessage);
        controller.teardown();
    };
}
export function createPageHeightBridgeController() {
    let vhCleanups = null;
    let vhForceRun = null;
    let pendingResponse;
    let pendingOrigin = "*";
    return {
        freezeVhUnits: (override) => {
            if (vhCleanups) {
                vhForceRun?.();
                return;
            }
            const referenceVhBase = resolveReferenceVhBase(override);
            vhCleanups = [];
            vhForceRun = startVhNeutralizer(referenceVhBase, vhCleanups);
        },
        // Target the requester's origin so the height isn't broadcast to anyone
        // who happens to embed us. Falls back to "*" when origin is unavailable
        // (jsdom default, sandboxed iframes with `null` origin).
        measurePageHeight: (origin, settleMs = DEFAULT_SETTLE_MS) => {
            pendingOrigin = origin;
            if (pendingResponse !== undefined)
                window.clearTimeout(pendingResponse);
            pendingResponse = window.setTimeout(() => {
                requestAnimationFrame(() => {
                    const height = measureContentHeight();
                    window.parent.postMessage({ type: "page-height-measured", height }, pendingOrigin);
                });
            }, settleMs);
        },
        teardown: () => {
            if (vhCleanups) {
                for (const c of vhCleanups)
                    c();
                vhCleanups = null;
                vhForceRun = null;
            }
            if (pendingResponse !== undefined)
                window.clearTimeout(pendingResponse);
        },
    };
}
function resolveReferenceVhBase(override) {
    if (override !== undefined)
        return override;
    const detected = window.innerHeight || 0;
    return detected >= MIN_VHBASE ? detected : FALLBACK_VHBASE;
}
// Caches keep work proportional to *new* CSS, not total CSS. <head> observer
// catches `<style>`/`<link>` adds; per-element observers catch HMR text edits.
// Returns a `forceRun` so the caller can re-trigger neutralization on later
// parent requests (idempotent — already-rewritten text is skipped via the
// processed sets).
function startVhNeutralizer(referenceVhBase, cleanups) {
    const VH_RE = /(\d+(?:\.\d+)?)vh\b/g;
    const processedStyles = new WeakSet();
    const processedSheets = new WeakMap();
    const watchedStyles = new WeakSet();
    const styleObservers = new Set();
    const neutralize = () => {
        const rewrite = (input) => input.replace(VH_RE, (_match, n) => `${((parseFloat(n) / 100) * referenceVhBase).toFixed(2)}px`);
        document.querySelectorAll("style").forEach((el) => {
            watchStyleEl(el);
            if (processedStyles.has(el))
                return;
            processedStyles.add(el);
            const text = el.textContent;
            if (!text || text.indexOf("vh") === -1)
                return;
            const next = rewrite(text);
            if (next !== text)
                el.textContent = next;
        });
        for (let i = 0; i < document.styleSheets.length; i++) {
            const sheet = document.styleSheets[i];
            if (!sheet)
                continue;
            let rules;
            try {
                rules = sheet.cssRules;
            }
            catch {
                continue; // CORS-protected
            }
            if (processedSheets.get(sheet) === rules.length)
                continue;
            rewriteVhInRules(rules, rewrite);
            processedSheets.set(sheet, rules.length);
        }
    };
    const debouncer = createDebouncer(neutralize, NEUTRALIZE_DEBOUNCE_MS);
    const debouncedNeutralize = debouncer.trigger;
    cleanups.push(debouncer.cancel);
    const watchStyleEl = (el) => {
        if (watchedStyles.has(el))
            return;
        watchedStyles.add(el);
        const obs = new MutationObserver(() => {
            processedStyles.delete(el);
            debouncedNeutralize();
        });
        obs.observe(el, { characterData: true, childList: true, subtree: true });
        styleObservers.add(obs);
    };
    cleanups.push(() => {
        for (const obs of styleObservers)
            obs.disconnect();
        styleObservers.clear();
    });
    debouncedNeutralize();
    if (document.readyState === "loading") {
        document.addEventListener("DOMContentLoaded", debouncedNeutralize);
        cleanups.push(() => document.removeEventListener("DOMContentLoaded", debouncedNeutralize));
    }
    window.addEventListener("load", debouncedNeutralize);
    cleanups.push(() => window.removeEventListener("load", debouncedNeutralize));
    const headObserver = new MutationObserver((mutations) => {
        for (const m of mutations) {
            if (containsStylesheetNode(m.addedNodes) || containsStylesheetNode(m.removedNodes)) {
                debouncedNeutralize();
                return;
            }
        }
    });
    cleanups.push(() => headObserver.disconnect());
    const attachHeadObserver = () => {
        if (document.head)
            headObserver.observe(document.head, { childList: true, subtree: false });
    };
    if (document.head) {
        attachHeadObserver();
    }
    else {
        document.addEventListener("DOMContentLoaded", attachHeadObserver);
        cleanups.push(() => document.removeEventListener("DOMContentLoaded", attachHeadObserver));
    }
    return debouncedNeutralize;
}
function rewriteVhInRules(rules, rewrite) {
    for (let i = 0; i < rules.length; i++) {
        const rule = rules[i];
        if (!rule)
            continue;
        if (rule.cssRules)
            rewriteVhInRules(rule.cssRules, rewrite);
        const style = rule.style;
        if (!style)
            continue;
        for (let j = 0; j < style.length; j++) {
            const prop = style[j];
            if (!prop)
                continue;
            const value = style.getPropertyValue(prop);
            if (!value || value.indexOf("vh") === -1)
                continue;
            const next = rewrite(value);
            if (next !== value)
                style.setProperty(prop, next, style.getPropertyPriority(prop));
        }
    }
}
function containsStylesheetNode(nodes) {
    for (let i = 0; i < nodes.length; i++) {
        const node = nodes[i];
        if (node instanceof HTMLStyleElement || node instanceof HTMLLinkElement)
            return true;
    }
    return false;
}
function measureContentHeight() {
    return Math.max(document.documentElement.scrollHeight, document.body?.scrollHeight ?? 0);
}
function createDebouncer(fn, delayMs) {
    let timer;
    return {
        trigger: () => {
            if (timer !== undefined)
                window.clearTimeout(timer);
            timer = window.setTimeout(fn, delayMs);
        },
        cancel: () => {
            if (timer !== undefined) {
                window.clearTimeout(timer);
                timer = undefined;
            }
        },
    };
}
