// @sentry/browser/build/npm/cjs/prod/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const feedbackAsync = require('./feedbackAsync.js');
const feedbackSync = require('./feedbackSync.js');
const browser = require('@sentry/core/browser');
const helpers = require('./helpers.js');
const client = require('./client.js');
const fetch = require('./transports/fetch.js');
const index = require('./profiling/index.js');
const stackParsers = require('./stack-parsers.js');
const eventbuilder = require('./eventbuilder.js');
const userfeedback = require('./userfeedback.js');
const sdk = require('./sdk.js');
const reportDialog = require('./report-dialog.js');
const breadcrumbs = require('./integrations/breadcrumbs.js');
const globalhandlers = require('./integrations/globalhandlers.js');
const httpcontext = require('./integrations/httpcontext.js');
const linkederrors = require('./integrations/linkederrors.js');
const browserapierrors = require('./integrations/browserapierrors.js');
const browsersession = require('./integrations/browsersession.js');
const lazyLoadIntegration = require('./utils/lazyLoadIntegration.js');
const reportingobserver = require('./integrations/reportingobserver.js');
const httpclient = require('./integrations/httpclient.js');
const contextlines = require('./integrations/contextlines.js');
const graphqlClient = require('./integrations/graphqlClient.js');
const viewHierarchy = require('./integrations/view-hierarchy.js');
const replay = require('@sentry-internal/replay');
const replayCanvas = require('@sentry-internal/replay-canvas');
const feedback = require('@sentry-internal/feedback');
const request = require('./tracing/request.js');
const browserTracingIntegration = require('./tracing/browserTracingIntegration.js');
const browserUtils = require('@sentry-internal/browser-utils');
const reportPageLoaded = require('./tracing/reportPageLoaded.js');
const setActiveSpan = require('./tracing/setActiveSpan.js');
const spanstreaming = require('./integrations/spanstreaming.js');
const offline = require('./transports/offline.js');
const integration$1 = require('./profiling/integration.js');
const spotlight = require('./integrations/spotlight.js');
const culturecontext = require('./integrations/culturecontext.js');
const integration$2 = require('./integrations/featureFlags/launchdarkly/integration.js');
const integration = require('./integrations/featureFlags/openfeature/integration.js');
const integration$5 = require('./integrations/featureFlags/unleash/integration.js');
const integration$3 = require('./integrations/featureFlags/growthbook/integration.js');
const integration$4 = require('./integrations/featureFlags/statsig/integration.js');
const diagnoseSdk = require('./diagnose-sdk.js');
const webWorker = require('./integrations/webWorker.js');



exports.feedbackAsyncIntegration = feedbackAsync.feedbackAsyncIntegration;
exports.feedbackIntegration = feedbackSync.feedbackSyncIntegration;
exports.feedbackSyncIntegration = feedbackSync.feedbackSyncIntegration;
exports.MULTIPLEXED_TRANSPORT_EXTRA_KEY = browser.MULTIPLEXED_TRANSPORT_EXTRA_KEY;
exports.SDK_VERSION = browser.SDK_VERSION;
exports.SEMANTIC_ATTRIBUTE_SENTRY_OP = browser.SEMANTIC_ATTRIBUTE_SENTRY_OP;
exports.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN = browser.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE = browser.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE = browser.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE;
exports.Scope = browser.Scope;
exports.addBreadcrumb = browser.addBreadcrumb;
exports.addEventProcessor = browser.addEventProcessor;
exports.addIntegration = browser.addIntegration;
exports.captureConsoleIntegration = browser.captureConsoleIntegration;
exports.captureEvent = browser.captureEvent;
exports.captureException = browser.captureException;
exports.captureFeedback = browser.captureFeedback;
exports.captureMessage = browser.captureMessage;
exports.captureSession = browser.captureSession;
exports.close = browser.close;
exports.consoleLoggingIntegration = browser.consoleLoggingIntegration;
exports.continueTrace = browser.continueTrace;
exports.createConsolaReporter = browser.createConsolaReporter;
exports.createLangChainCallbackHandler = browser.createLangChainCallbackHandler;
exports.createTransport = browser.createTransport;
exports.dedupeIntegration = browser.dedupeIntegration;
exports.endSession = browser.endSession;
exports.eventFiltersIntegration = browser.eventFiltersIntegration;
exports.extraErrorDataIntegration = browser.extraErrorDataIntegration;
exports.featureFlagsIntegration = browser.featureFlagsIntegration;
exports.flush = browser.flush;
exports.functionToStringIntegration = browser.functionToStringIntegration;
exports.getActiveSpan = browser.getActiveSpan;
exports.getClient = browser.getClient;
exports.getCurrentScope = browser.getCurrentScope;
exports.getGlobalScope = browser.getGlobalScope;
exports.getIsolationScope = browser.getIsolationScope;
exports.getRootSpan = browser.getRootSpan;
exports.getSpanDescendants = browser.getSpanDescendants;
exports.getSpanStatusFromHttpCode = browser.getSpanStatusFromHttpCode;
exports.getTraceData = browser.getTraceData;
exports.inboundFiltersIntegration = browser.inboundFiltersIntegration;
exports.instrumentAnthropicAiClient = browser.instrumentAnthropicAiClient;
exports.instrumentCreateReactAgent = browser.instrumentCreateReactAgent;
exports.instrumentGoogleGenAIClient = browser.instrumentGoogleGenAIClient;
exports.instrumentLangChainEmbeddings = browser.instrumentLangChainEmbeddings;
exports.instrumentLangGraph = browser.instrumentLangGraph;
exports.instrumentOpenAiClient = browser.instrumentOpenAiClient;
exports.instrumentSupabaseClient = browser.instrumentSupabaseClient;
exports.isEnabled = browser.isEnabled;
exports.isInitialized = browser.isInitialized;
exports.lastEventId = browser.lastEventId;
exports.logger = browser.logger;
exports.makeMultiplexedTransport = browser.makeMultiplexedTransport;
exports.metrics = browser.metrics;
exports.moduleMetadataIntegration = browser.moduleMetadataIntegration;
exports.parameterize = browser.parameterize;
exports.registerSpanErrorInstrumentation = browser.registerSpanErrorInstrumentation;
exports.rewriteFramesIntegration = browser.rewriteFramesIntegration;
exports.setContext = browser.setContext;
exports.setConversationId = browser.setConversationId;
exports.setCurrentClient = browser.setCurrentClient;
exports.setExtra = browser.setExtra;
exports.setExtras = browser.setExtras;
exports.setHttpStatus = browser.setHttpStatus;
exports.setMeasurement = browser.setMeasurement;
exports.setTag = browser.setTag;
exports.setTags = browser.setTags;
exports.setUser = browser.setUser;
exports.spanToBaggageHeader = browser.spanToBaggageHeader;
exports.spanToJSON = browser.spanToJSON;
exports.spanToTraceHeader = browser.spanToTraceHeader;
exports.startInactiveSpan = browser.startInactiveSpan;
exports.startNewTrace = browser.startNewTrace;
exports.startSession = browser.startSession;
exports.startSpan = browser.startSpan;
exports.startSpanManual = browser.startSpanManual;
exports.supabaseIntegration = browser.supabaseIntegration;
exports.suppressTracing = browser.suppressTracing;
exports.thirdPartyErrorFilterIntegration = browser.thirdPartyErrorFilterIntegration;
exports.updateSpanName = browser.updateSpanName;
exports.withActiveSpan = browser.withActiveSpan;
exports.withIsolationScope = browser.withIsolationScope;
exports.withScope = browser.withScope;
exports.withStreamedSpan = browser.withStreamedSpan;
exports.zodErrorsIntegration = browser.zodErrorsIntegration;
exports.WINDOW = helpers.WINDOW;
exports.BrowserClient = client.BrowserClient;
exports.makeFetchTransport = fetch.makeFetchTransport;
exports.uiProfiler = index.uiProfiler;
exports.chromeStackLineParser = stackParsers.chromeStackLineParser;
exports.defaultStackLineParsers = stackParsers.defaultStackLineParsers;
exports.defaultStackParser = stackParsers.defaultStackParser;
exports.geckoStackLineParser = stackParsers.geckoStackLineParser;
exports.opera10StackLineParser = stackParsers.opera10StackLineParser;
exports.opera11StackLineParser = stackParsers.opera11StackLineParser;
exports.winjsStackLineParser = stackParsers.winjsStackLineParser;
exports.eventFromException = eventbuilder.eventFromException;
exports.eventFromMessage = eventbuilder.eventFromMessage;
exports.exceptionFromError = eventbuilder.exceptionFromError;
exports.createUserFeedbackEnvelope = userfeedback.createUserFeedbackEnvelope;
exports.forceLoad = sdk.forceLoad;
exports.getDefaultIntegrations = sdk.getDefaultIntegrations;
exports.init = sdk.init;
exports.onLoad = sdk.onLoad;
exports.showReportDialog = reportDialog.showReportDialog;
exports.breadcrumbsIntegration = breadcrumbs.breadcrumbsIntegration;
exports.globalHandlersIntegration = globalhandlers.globalHandlersIntegration;
exports.httpContextIntegration = httpcontext.httpContextIntegration;
exports.linkedErrorsIntegration = linkederrors.linkedErrorsIntegration;
exports.browserApiErrorsIntegration = browserapierrors.browserApiErrorsIntegration;
exports.browserSessionIntegration = browsersession.browserSessionIntegration;
exports.lazyLoadIntegration = lazyLoadIntegration.lazyLoadIntegration;
exports.reportingObserverIntegration = reportingobserver.reportingObserverIntegration;
exports.httpClientIntegration = httpclient.httpClientIntegration;
exports.contextLinesIntegration = contextlines.contextLinesIntegration;
exports.graphqlClientIntegration = graphqlClient.graphqlClientIntegration;
exports.viewHierarchyIntegration = viewHierarchy.viewHierarchyIntegration;
exports.getReplay = replay.getReplay;
exports.replayIntegration = replay.replayIntegration;
exports.replayCanvasIntegration = replayCanvas.replayCanvasIntegration;
exports.getFeedback = feedback.getFeedback;
exports.sendFeedback = feedback.sendFeedback;
exports.defaultRequestInstrumentationOptions = request.defaultRequestInstrumentationOptions;
exports.instrumentOutgoingRequests = request.instrumentOutgoingRequests;
exports.browserTracingIntegration = browserTracingIntegration.browserTracingIntegration;
exports.isBotUserAgent = browserTracingIntegration.isBotUserAgent;
exports.startBrowserTracingNavigationSpan = browserTracingIntegration.startBrowserTracingNavigationSpan;
exports.startBrowserTracingPageLoadSpan = browserTracingIntegration.startBrowserTracingPageLoadSpan;
exports.elementTimingIntegration = browserUtils.elementTimingIntegration;
exports.reportPageLoaded = reportPageLoaded.reportPageLoaded;
exports.setActiveSpanInBrowser = setActiveSpan.setActiveSpanInBrowser;
exports.spanStreamingIntegration = spanstreaming.spanStreamingIntegration;
exports.makeBrowserOfflineTransport = offline.makeBrowserOfflineTransport;
exports.browserProfilingIntegration = integration$1.browserProfilingIntegration;
exports.spotlightBrowserIntegration = spotlight.spotlightBrowserIntegration;
exports.cultureContextIntegration = culturecontext.cultureContextIntegration;
exports.buildLaunchDarklyFlagUsedHandler = integration$2.buildLaunchDarklyFlagUsedHandler;
exports.launchDarklyIntegration = integration$2.launchDarklyIntegration;
exports.OpenFeatureIntegrationHook = integration.OpenFeatureIntegrationHook;
exports.openFeatureIntegration = integration.openFeatureIntegration;
exports.unleashIntegration = integration$5.unleashIntegration;
exports.growthbookIntegration = integration$3.growthbookIntegration;
exports.statsigIntegration = integration$4.statsigIntegration;
exports.diagnoseSdkConnectivity = diagnoseSdk.diagnoseSdkConnectivity;
exports.registerWebWorker = webWorker.registerWebWorker;
exports.webWorkerIntegration = webWorker.webWorkerIntegration;
